src/flag_gems/ops/nllloss.py

```diff
@@ -0,0 +1,311 @@
+        )
+        tl.store(inp_grad_ptrs, inp_grad.to(tl.float32), mask=(inp_mask & ignore_mask))
+
+
+# Negative Log Likelihood Loss (NLLLoss)
+#
+# This loss function is used for training classification problems with C classes.
+#
+# Parameters:
+# - input (Tensor):
+#   - Expected to contain log-probabilities for each class.
+#   - Shape can be either:
+#     - (minibatch, C) for standard classification tasks.
+#     - (minibatch, C, d1, d2, ..., dK) for K-dimensional inputs (e.g., per-pixel loss for 2D images).
+#
+# - target (Tensor):
+#   - Should contain class indices in the range [0, C-1].
+#   - If ignore_index is specified, this index can be outside the class range
+#       and will be ignored in the loss computation.
+#
+# - weight (1D Tensor, optional):
+#   - Assigns weight to each class, useful for unbalanced datasets.
+#
+# Reduction modes:
+# - 'none': returns per-sample loss (shape: (N,)).
+# - 'mean' (default): computes the mean of the weighted losses.
+# - 'sum': computes the sum of the weighted losses.
+#
+# Mathematical description:
+# - Unreduced loss:
+#   l_n = -w_y_n * x_n, where w_c = weight[c] * 1{c != ignore_index}.
+# - Reduced loss (depending on the specified reduction mode):
+#   - mean: ℓ(x, y) = (1/N) * Σ(w_y_n * l_n)
+#   - sum: ℓ(x, y) = Σ(l_n)
+class NegativeLogLikeLoss(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, inp, target, weight, reduction, ignore_index):
+        logging.debug("GEMS NLLLoss FWD")
+        shape = list(inp.shape)
+        dim = inp.ndim
+        N = 1 if dim == 1 else shape[0]
+        C = shape[0] if dim == 1 else shape[1]
+        D = inp.numel() // N // C
+        axis = 0 if dim == 1 else 1
+        del shape[axis]
+
+        assert ((i >= 0 and i < C) for i in target), "Target is out of bounds"
+        assert list(target.shape) == shape, "Invalid target size"
+        assert inp.ndim >= 1, "Invalid input ndim"
+
+        if weight is None:
+            weight = torch.ones(
+                (C),
+                dtype=inp.dtype,
+                device=inp.device,
+            )
+
+        inp = inp.contiguous()
+        tgt = target.contiguous()
+        w = weight.contiguous()
+        out = torch.zeros(shape, dtype=torch.float32, device=inp.device)
+        w_tgt = torch.zeros(shape, dtype=torch.float32, device=inp.device)
+
+        if inp.ndim > 2:
+            grid = lambda meta: (N, triton.cdiv(D, meta["BLOCK_D"]))
+            with torch.cuda.device(inp.device):
+                nll_loss_multi_fwd_kernel[grid](
+                    inp, tgt, w, w_tgt, out, ignore_index, N, C, D
+                )
+        else:
+            grid = lambda meta: (triton.cdiv(N, meta["BLOCK_N"]),)
+            with torch.cuda.device(inp.device):
+                nll_loss_2d_fwd_kernel[grid](
+                    inp, tgt, w, w_tgt, out, ignore_index, N, C
+                )
+
+        ctx.save_for_backward(inp, tgt, w)
```

Comment:
only saving tensors and variables when input requires gradient might decrease the cost.